src/overlay/test/OverlayTests.cpp

```diff
@@ -1139,6 +1139,139 @@ TEST_CASE("inbounds nodes can be promoted to ouboundvalid",
     simulation->crankForAtLeast(std::chrono::seconds{3}, true);
 }
 
+TEST_CASE("outbound message ordering", "[overlay]")
+{
+    std::multiset<Peer::QueuedOutboundMessage, Peer::OutboundMsgCmp> testSet;
+
+    StellarMessage scpMsg1;
+    StellarMessage scpMsg2;
+    scpMsg1.type(SCP_MESSAGE);
+    scpMsg2 = scpMsg1;
+
+    auto time = VirtualClock::time_point();
+    auto time1 = time + std::chrono::seconds(1);
+    auto time2 = time1 + std::chrono::seconds(1);
+    auto time3 = time2 + std::chrono::seconds(1);
+    auto time4 = time3 + std::chrono::seconds(1);
+
+    StellarMessage tx1;
+    StellarMessage tx2;
+    tx1.type(TRANSACTION);
+    tx2 = tx1;
+
+    StellarMessage pingResp;
+    pingResp.type(DONT_HAVE);
+    pingResp.dontHave().type = SCP_QUORUMSET;
+
+    testSet.insert(Peer::QueuedOutboundMessage{tx1, time});
+    testSet.insert(Peer::QueuedOutboundMessage{tx2, time1});
+    testSet.insert(Peer::QueuedOutboundMessage{scpMsg1, time2});
+    testSet.insert(Peer::QueuedOutboundMessage{pingResp, time3});
+    testSet.insert(Peer::QueuedOutboundMessage{scpMsg2, time4});
+
+    auto it = testSet.cbegin();
+    REQUIRE(it->mMessage.type() == SCP_MESSAGE);
+    REQUIRE(it->mTimeEmplaced == time2);
+
+    ++it;
+
+    REQUIRE(it->mMessage.type() == DONT_HAVE);
+    REQUIRE(it->mTimeEmplaced == time3);
+
+    ++it;
+
+    REQUIRE(it->mMessage.type() == SCP_MESSAGE);
+    REQUIRE(it->mTimeEmplaced == time4);
+
+    ++it;
+
+    REQUIRE(it->mMessage.type() == TRANSACTION);
+    REQUIRE(it->mTimeEmplaced == time);
+
+    ++it;
+
+    REQUIRE(it->mMessage.type() == TRANSACTION);
+    REQUIRE(it->mTimeEmplaced == time1);
+}
+
+TEST_CASE("overlay flow control", "[overlay][flowcontrol]")
```

Comment:
missing tests:
* old vs new (new wants to enable flow control, but old doesn't support it)
* handshake tests
  * various combinations for "send_more" (peer disconnects when receiving it at the wrong time with illegal values like 0 or very large numbers)
  * check that nodes agree on flow control being activated or not
* tests that verify that we don't read too many things at once
* tests that check send_more contract respected (adjust flood capacity on the fly, drop peers that don't respect contract)